### DISCUSSION TITLE
Graph cache update behavior

### DISCUSSION THREAD

**Discussion by focux** [urql-graphql/urql#2427] [Q&A] [answered]
I spent some days trying to debug why some cache updates coming from an optimistic update were not working and I just figured it out and would like to know if it's the right behavior or if it's a bug. So, I'm working on a chat, and every time a user sends a message, I optimistically update the cache like this:

```typescript
{
    message: {
        id: "123",
        text: "Hello world",
        user: {
            id: "1",
            displayName: "Lisa",
            __typename: "User",
        },
        __typename: "ChatMessage",
    },
    __typename: "ChatMessageResponse",
}
```

Here comes the issue, when the user, in this case, `User:1` is cached and has the `displayName` property on the cached object, then the updates work, but if somehow, the cached version doesn't have the `displayName` property, then it doesn't update the cache or log/throw any error.

I would expect the update to succeed even if the `displayName` is not cached. I mean, can't we just add the missing property that the optimistic update returned to the cached object?

Is this really how it should work or is it a bug?

**Comment by kitten** [comment, marked as answer]
This really depends on your queries.

Queries are altered by optimistic updates, yes, but not directly. Instead, all results from the API (and optimistic mutations specifically) update the internal data store. In other words, there's no direct path from your optimistic update to the queries.
Rather, when the cache changes then the appropriate queries that contain updated data are "reexecuted". This is the same as if you would've just triggered the query, e.g. by switching SPA pages.

I suspect what's happened is that your optimistic update causes a query to update which needs this `dispkayName`. There's no sure fire way to guarantee that this yields complete data as you say. Your optimistic mutation's selection set (and optimistic data) may contain less data than what a query somewhere else depends on.

What then happens is that if you trigger the optimistic update, Graphcache sees a query that depends on this data and refreshes it. If this query then contains missing data it's irrelevant that an optimistic update triggered it. It simply contains missing data i.e a cache miss and must hence be refetched from the API. this can only happen however once the optimistic mutation has finished to prevent your optimistic update from being overridden/deleted.

One way to prevent this is to use Schema Awareness which may allow you to receive and display partial results: https://formidable.com/open-source/urql/docs/graphcache/schema-awareness/

We also planned to extend the `populateExchange` to improve your mutation behaviour. But optimistic mutations are ultimately a special case.

**Reply by focux** [reply]
So, if I understand right, if I use schema awareness then my optimistic updates can work even with partial data? Because I'm already using that and that's not the case, at least on that specific example.

Coming back to the example I used above, I would expect the optimistic update to execute even though the cached user doesn't have the `displayName` property because I'm passing a `displayName` through the optimistic update payload. So, why just not merge the optimistic update payload with the already cached objects.

Also, is there a way to console log a meaningful warning when the optimistic update didn't execute due to missing fields?

And thank you so much for taking the time to answer me!

**Reply by erquhart** [reply]
@kitten I'm curious on this as well (dealing with a similar issue) - the developer is saying that the optimistic update has a field that the cached version does not, so this isn't a partial update, it's a complete update with an additional field that just isn't in the cache yet. I'm finding that these subtle differences are not only blocking optimistic updates, but also the subsequent network update in some cases.
